### DISCUSSION TITLE
Announcing GitHub Copilot Labs

### DISCUSSION THREAD

**Discussion by irealva** [community/community#8308] [Copilot Conversations]
**⚠️ UPDATE March 30 ⚠️**: The new GitHub Copilot Labs extension is [HERE](https://marketplace.visualstudio.com/items?itemName=GitHub.copilot-labs). And our new discussion forum is [HERE](https://github.com/github/feedback/discussions/13901).

**⚠️ UPDATE March 28 ⚠️**: We are deprecating Labs from the Copilot Nightly extension starting tomorrow! But don't worry, these features are not going away. Labs will be available as a separate extension we'll release very soon. Thanks to everyone who has been an early user!

Greetings!

You may have noticed some cool new functionality in [recent nightlies of GitHub Copilot](https://marketplace.visualstudio.com/items?itemName=GitHub.copilot-nightly) for Visual Studio Code! We’ve added the Labs sidebar, a home for experiments we’re building with the same technologies that power GitHub Copilot.

The first application is Explain. Select some code, and ask GitHub Copilot to explain that code to you in plain language:

![GIF depicting the Copilot Labs Explain Flow](https://user-images.githubusercontent.com/22723/144143451-dc9fd535-ef22-4077-8ffe-dbbe5d322805.gif)

### Create custom prompts

We provide a few preset prompts to get you started: three that explain what a particular block of code does, and another that generates example code for calling a function.

You can customize the prompt and stop sequence of a query in order to come up with new applications that use [Codex](https://openai.com/blog/openai-codex/) to interpret code. Creating these can feel like more of an art than a science! Small changes in the formulation of the prompt and stop sequence can produce very different results. The three different “explain” examples showcase strategies that tend to produce useful responses from the model, but this is uncharted territory! We’re excited to see what you use this for.

These articles on [prompt design](https://beta.openai.com/docs/guides/completion/prompt-design) and [stop sequences](https://help.openai.com/en/articles/5072263-how-do-i-use-stop-sequences) are a great place to start if you want to craft your own presets.

[Please share feedback](https://github.com/github/feedback/discussions/categories/copilot-feedback) on GitHub Copilot Labs! Have you found a prompt and stop sequence combination that works well, or did you generate something particularly helpful? Let us know so we improve the sidebar, and implement more useful presets and features over time.

— GitHub Next

**Comment by minimaxir** [comment]
I mean, it's not wrong.

<img width="439" alt="Screen Shot 2021-12-01 at 8 39 20 AM" src="https://user-images.githubusercontent.com/2179708/144275325-1ae3fd35-4cf2-43fb-ad34-1d5f99832d75.png">

Ok, ok, that's cheating. But for fun I put some of my [galaxy-brain AI generated code examples](https://minimaxir.com/2021/06/gpt-j-6b/) pre-Cortex to see what happens.

<img width="434" alt="Screen Shot 2021-12-01 at 8 49 52 AM" src="https://user-images.githubusercontent.com/2179708/144277323-9ec4bec6-f068-47c2-9833-b27f9616f81f.png">

<img width="437" alt="Screen Shot 2021-12-01 at 8 53 17 AM" src="https://user-images.githubusercontent.com/2179708/144278226-33f91e15-1f54-4372-8d0f-311df71dda57.png">

<img width="426" alt="Screen Shot 2021-12-01 at 8 54 34 AM" src="https://user-images.githubusercontent.com/2179708/144278253-d3792fe8-b3ed-4cd2-87f5-72ec43627009.png">

<img width="427" alt="Screen Shot 2021-12-01 at 8 55 28 AM" src="https://user-images.githubusercontent.com/2179708/144278278-822683a8-7696-4878-9d21-1476977da886.png">

tbh, better than expected.

**Reply by idan** [reply]
Wow, never thought to use this to explain regexes in plain english, mindblown.gif

**Reply by jonathankoren** [reply]
The explanation is wrong though. It's four alphanums, a dollar sign, and then three more alphanums.

**Reply by mathmul** [reply]
Not wrong, incomplete.

**Reply by Selena20191** [reply]
> The explanation is wrong though. It's four alphanums, a dollar sign, and then three more alphanums.

**Comment by TimDaub** [comment]


**Reply by ganeshkrishnan1** [reply]
This is not the place to discuss it.

**Reply by TimDaub** [reply]
makes sense

**Reply by Selena20191** [reply]
> makes sense
W

**Comment by Milo123459** [comment]
When is the stable version containing explain coming out?

**Reply by idan** [reply]
Hi @Milo123459! We are keeping this in Nightlies for now to test, and I imagine that we'll promote it into mainline builds early in the new year. No commitments yet in terms of timeline.

**Comment by quackduck** [comment]
Amazing!

Do you think the JetBrains release of new features will lag behind the VS Code release? (that makes sense to me)

Maybe around a month or so?

**Reply by idan** [reply]
Labs is VS Code-only for now!

**Reply by quackduck** [reply]
what do you think the lag may be?

**Reply by Selena20191** [reply]
👍

**Comment by bdougie** [comment]
Tested this out earlier today and it was impressive. 

A suggestion for a future enhancement: Adding function headers output, i.e. like in [JavaScript](https://www.codexpedia.com/javascript/javascript-function-and-class-header-documentation/) would be amazing. I work on a function and then "Ask Copilot" for the headers.

**Reply by irealva** [reply]
Thanks for the suggestion!

**Comment by datapolitical** [comment]
This looks awesome! What is the typical wait like at present to join the beta?

**Reply by MattStopa** [reply]
3 eternities.

**Reply by datapolitical** [reply]
Well that sounds promising

**Reply by aidantorrence** [reply]
it's 24 hours

**Reply by Selena20191** [reply]
👍

**Comment by AashJ** [comment]
Are there plans for an API?

**Reply by irealva** [reply]
Not for now, sorry.

**Comment by TarunavBA** [comment]
Will people who are in problems as in #7716 Or related will be able to use this even without copilot?

**Reply by adarshaacharya** [reply]
@TarunavBA i think you should delete github copilot and install github copilot nightly.

**Comment by CaelumF** [comment]
This seems to go further in providing unrestricted access to the underlying Codex API. Has there been any thoughts about the potential for abuse, for example non-code prompts like the generation of propaganda and fabricated news? On [OpenAI's use case guidelines](https://beta.openai.com/docs/usage-guidelines/use-case-guidelines) it disallows serving open-ended inputs, and my understanding was that this is for all of the APIs including Codex? Are there any filters for this feature?

**Comment by MulleDK19** [comment]
Tried a few things. Not really helpful so far since it's hardcoded to cut off after one line!?
No matter what, it generates only one line, then stops, so I'm only getting very incomplete outputs.

**Reply by CaelumF** [reply]
What are your stops sequences ? Perhaps if none are specified then newline is the default

**Reply by irealva** [reply]
Was going to write the same, some combinations of prompts and stop sequences generate no response. Or post an example so we can take a look.

**Reply by MulleDK19** [reply]
Stops make no difference, with or without, I get one line back, always. No matter what stop I put, it just generates one line ending in whatever stop I've specified.

![image](https://user-images.githubusercontent.com/3169809/144427232-eadce5de-73b9-4ee7-9caf-9cf427da3cf7.png)

**Reply by MulleDK19** [reply]
![image](https://user-images.githubusercontent.com/3169809/144427313-e3401bfc-f9b1-40f6-b1a8-1934b814870d.png)

**Reply by irealva** [reply]
Will try to replicate

**Reply by irealva** [reply]
I believe your first example is a csharp file. Not sure why but in this case if you create the prompt without `1.` it gives you a better answer. Again this is where there's a ton of variation in prompt building between languages and code samples. We're learning along with you! 

I tried:

- `/* The code above does the following: ` space at the end is important!
- `/* Here is the explanation for the code above: ` space at the end is important!

<img width="298" alt="Screen Shot 2021-12-10 at 11 00 12 AM" src="https://user-images.githubusercontent.com/8978670/145603832-8857d781-e573-4b72-bb45-2f6f6855d7dd.png">

<img width="306" alt="Screen Shot 2021-12-10 at 11 00 33 AM" src="https://user-images.githubusercontent.com/8978670/145603851-e8bc7238-40b7-40d1-be5c-552515f06bee.png">

**Comment by ldorigo** [comment]
Neat! Love that you allow making prompts - hopefully soon it will be possible to design and save prompts and call them with special commands and keybindings. Specialized prompts for generating docstrings, normalizing code conventions, adding type annotations, ... so many possibilities.

Regarding code explanation, you can take a look at a post I made on Codex line-by-line explanation that works fantastically: https://community.openai.com/t/best-line-by-line-code-explanation-prompt-i-found-so-far/12272/6

EDIT: Just tried it, and it looks like it doesn't work for all languages? I tried it on a shell script and it does not react to when I highlight code.

**Reply by irealva** [reply]
We'll take a look! Should work for all languages...

**Reply by irealva** [reply]
Hi @ldorigo thanks for reporting this and for your patience. We fixed the language problem and it should work on the next nightly release for shell scripts.

**Comment by Milo123459** [comment]
Not sure if this was already said, but maybe ignoring a bulk of indentation would be nice?
<img width="121" alt="image" src="https://user-images.githubusercontent.com/50248166/144464898-129fb0ec-2ea2-4d6b-9ce0-4c0d91332506.png">
 For example, see image above - maybe it could remove those tabs, but keep them when they open up into a callback or something.

**Comment by aidantorrence** [comment]
Took me like 15 minutes to figure out how to open the Labs sidebar.  Would be nice if there was better explanation.

**Reply by masayahingpuno** [reply]
> Took me like 15 minutes to figure out how to open the Labs sidebar. Would be nice if there was better explanation.

Hahaha me too!

**Reply by aidantorrence** [reply]
Here's an explanation for anyone confused:

1. Install Github Copilot Nightly and uninstall Github Copilot
2. There is a Github Copilot icon on the side, click it and that opens up the labs sidebar.


https://user-images.githubusercontent.com/26528573/144504040-9f1ace43-3833-47a5-85f2-a6873c429736.mov

**Reply by masayahingpuno** [reply]
You can also type Ctrl + Shift + P then search for Copilot Labs then create your own keyboard shortcut

**Reply by irealva** [reply]
Thanks for sharing the video @aidantorrence!

**Comment by aidantorrence** [comment]
Seems to only go to a certain length of explanation.  So if it's a larger snippet, it will often end prematurely before explaining everything.

![image](https://user-images.githubusercontent.com/26528573/144486756-a9a08f5b-9b7c-4ee4-9135-f42dfde06f23.png)

**Reply by aidantorrence** [reply]
I like:

```
/* above code in plain english:

1.
```

**Reply by irealva** [reply]
We'll try to figure out how to not cut these off prematurely.

**Reply by irealva** [reply]
Hi @aidantorrence this should be fixed in the next nightly build. The explanations shouldn't cut off prematurely anymore.

**Comment by ShauryaAg** [comment]
**regex** to explain: `/(?<=@)(\w+)/g`

**Result**:
![copilot](https://user-images.githubusercontent.com/31778302/144569917-24d40f74-bf28-4b04-87b1-1f02c5cb512c.jpeg)

The explanation of the above **regex** says it matches the word against "hello" (which it does not)
The above **regex** only looks for words after `@` symbol, for example: `@ShauryaAg`

PS: Tried to replicate, but couldn't.

**Reply by irealva** [reply]
The model is certainly not perfect! You'd be helping improve it — or at least helping us understand where it performs poorly — by submitting feedback with the buttons that show up after the code explanation.

**Reply by MeltingObsidian** [reply]
It was just giving an example by testing it against the string "hello" and explaining the outcome.

**Comment by aidantorrence** [comment]
doesn't seem to work in all instances of vscode(or all languages).  open up a new react app and it works, open a new flutter app and nothing shows up

![image](https://user-images.githubusercontent.com/26528573/145249681-90feb807-2dba-4d23-8226-69935c7564a0.png)

**Reply by irealva** [reply]
Hi @aidantorrence  thanks for reporting this and for your patience. We fixed the language problem and it should work on the next nightly release for your .dart files.

**Reply by datapolitical** [reply]
I'd love to help debug this but I don't have copilot

**Reply by aidantorrence** [reply]
@irealva awesome thanks!

**Reply by veb** [reply]
I have noticed when Copilot doesn't work, if you switch between say React JS then back to Babel JS, it'll work. Seems it needs a bit of a poke sometimes in order to work (especially labs)

**Reply by irealva** [reply]
Hey @veb could you share a short video of what you're seeing if you have time?

**Reply by veb** [reply]
@irealva but of course as soon as I start to record it doesn't replicate. heh. I am running both vscode insiders and copilot nightly so it could be fixed.  👀

What I had to do when I saw:
![image](https://user-images.githubusercontent.com/376245/146454852-f039df99-c9c7-4ab1-b2fa-77c41dd53544.png)

and then highlighted some text it wouldn't change until I switched the language in vscode and that would then work:
![image](https://user-images.githubusercontent.com/376245/146454995-393bce7c-3256-4c55-80d2-28b06f265392.png)

If it happens again I'll try and make sure to get some proof. I mainly posted it just to help others if they had come across the first screenshot not changing when highlighting a block of code. 🙃

**Reply by irealva** [reply]
We might have fixed it when we changed how we're parsing some of the languages for the sidebar. Just let me know if it happens again :)

**Comment by Jim-Salmons** [comment]
Irene @irealva (& Ryan @ryanjsalva, too),

You may be aware of my efforts to develop a support program and research study of [#GitHubCopilot for #DisabledDevelopers](https://medium.com/github-copilot-for-disabled-developers).

I have been in communication with the leadership team of the EU-based [Time Machine Project/Organization](https://www.timemachine.eu/) exploring their interest in collaborating with GitHub, Microsoft, and SciStarter.org on this proposed program. In reply, [Andreas Maier](https://lme.tf.fau.de/person/maier/)--one of my mentor/collaborators and Director of the Pattern Recognition Lab at FAU in Germany--offered to introduce me to a colleague working on support programs and technology for the visually-impaired. 

It immediately occurred to me that the Lab's Explain application--together with the built-in accessibility features of modern operating systems, like screen-reading speech generation—could be a useful innovation in support of visually-impaired developers.

When I make a connection with this FAU researcher, I would like to engage in an exploratory conversation with you and the Next group to see if there is interest in incorporating visual impairment support to the proposed **Copilot for Disabled Developers** research and support program agenda.

Happy Healthy Holiday Vibes from Colorado,
  Jim

**Reply by irealva** [reply]
Thanks for your note @Jim-Salmons. This is certainly an important and interesting application for Copilot. For now what I can say is I'll bring it up with the team. I'm not sure we're setup properly yet to support a program like this.

**Comment by Anselmoo** [comment]
> It is nice to get explained and showed example code! It really makes own code more reliable

I am wondering if it will be possible to directly communicate with Copilot similar to `HOWDOI` by [https://github.com/gleitz/howdoi](https://github.com/gleitz/howdoi)

`HOWDOI` is sometimes helpful, if you are still in a conceptual phase of coding.

## Example

```shell
❯ howdoi open json python tensorflow
```

#### Teriminal-Output

```python
import json
import numpy as np
import tensorflow as tf

def get_bbox(str):
    obj = json.loads(str.decode('utf-8'))
    bbox = obj['bounding_box']
    return np.array([bbox['x'], bbox['y'], bbox['height'], bbox['width']], dtype='f')

def get_multiple_bboxes(str):
    return [[get_bbox(x) for x in str]]

raw = tf.placeholder(tf.string, [None])
[parsed] = tf.py_func(get_multiple_bboxes, [raw], [tf.float32])
```

In this regard, it will be sometimes helpful if CoPilot suggest _Code_ **PLUS** _recommend essential packages_ in Python

**Comment by AbhimanyuAryan** [comment]
Julia suggestions can do better. I don't think co-pilot ever suggested me correct Julia. It throws `Ruby` and `Python` while writing `Julia`. However works really well with `Javascript`. Training the model on Julia SO, Julia Discourse, more Julia repos should help maybe?

**Comment by subercui** [comment]
Just curious, is this based on a newly fine-tuned codex? Or with prompt tuning?:smile:

**Reply by ryanjsalva** [reply]
We're constantly tweaking the experience. Model tuning takes a long time, so the smaller and more incremental changes you notice are more likely produced through a combination of prompt tuning and/or changes to our ranking+sorting algorithms (both of which can be implemented more efficiently).

**Comment by tfriedel** [comment]
I got some decent results by adding this prompt after a function:
```
# The above function can be rewritten as:
```
Example:
```
def f(n):  
    if n > 0:
        yield n
        
        for m in f(n - 1):
            yield m

# The above function can be rewritten as:
```
copilot:
```
# def f(n):
#     if n > 0:
#         yield n
#         yield from f(n - 1)
# 
```

**Comment by SiddharthShyniben** [comment]
This is cool! Waiting for vim support ;-;

**Reply by laoshaw** [reply]
yes please, need vim support

**Comment by irealva** [comment]
**⚠️ UPDATE March 28 ⚠️**: We are deprecating Labs from the Copilot Nightly extension starting tomorrow! But don't worry, these features are not going away. Labs will be available as a separate extension we'll release very soon. Thanks to everyone who has been an early user!

**Reply by MeltingObsidian** [reply]
We will eagerly wait for the new Labs extension. By the way, the Nightly extension will still be active, i.e., till the release of the new extension, its functionality will not be disabled, right? Or do we have to switch to the original Github Copilot extension?

**Reply by irealva** [reply]
Good question! So the Nightly extension will otherwise remain exactly the same. The only thing we're pulling out to a separate extension is the sidebar part (aka. the part we're calling Labs with all these early/experimental features).

**Reply by MeltingObsidian** [reply]
So, the sidebar will be removed with the next update to the Copilot Nightly extension. That's it?

**Reply by irealva** [reply]
Correct, you got it.

**Comment by irealva** [comment]
**⚠️ UPDATE March 30 ⚠️**: The new GitHub Copilot Labs extension is [HERE](https://marketplace.visualstudio.com/items?itemName=GitHub.copilot-labs). And our new discussion forum is [HERE](https://github.com/github/feedback/discussions/13901).

**Comment by liby** [comment]
@irealva 
It's the same code, but the explanation doesn't seem as detailed and useful as it used to be.
<img width="1225" alt="image" src="https://user-images.githubusercontent.com/38807139/167242226-5a99224c-4351-4a87-b53c-9446801cbdec.png">

**Comment by conioh** [comment]
I let Copilot write some code.
Copilot labs has a hard time understanding what Copilot prime wrote:

![image](https://user-images.githubusercontent.com/10606081/167464018-0feb53b2-247a-4698-be5c-b209f690d2ed.png)


![image](https://user-images.githubusercontent.com/10606081/167464179-e62fbeb6-4267-46ac-9b58-33729e1ee04c.png)

🤷

**Reply by veb** [reply]
Hey mate, I would suggest creating a new thread over on https://github.com/github/feedback/discussions and simply linking this one. 🙃

**Comment by xShazzy** [comment]
Will it be available in JetBrains tools?

**Reply by idan** [reply]
Hey! Sorry, we have no plans to implement labs for any editor outside of VS Code at this time.

**Reply by cantonalex** [reply]
Any updates on this?

**Reply by kvenn** [reply]
Yeah my team would benefit greatly from this on IntelliJ! Love the work ya'll are doing <3

**Reply by MCKevmeister** [reply]
> Hey! Sorry, we have no plans to implement labs for any editor outside of VS Code at this time.

Why not? It seems like a large portion of the user base uses copilot outside of vs code. Is it just that there would be no/less access to the telemetry of vs code? Is it that Microsoft just prefer everyone use vs code and wants to force users back into their ecosystem?

Not opening the tool up seems like a very unusual choice.

**Reply by dobsonddev** [reply]
The upstream, end-answer very much seems to be this:

> Microsoft just prefer everyone use vs code and wants to force users back into their ecosystem

*mr crabs* money money money money

Given the absurd market share that M has over dev solutions, it would only be true to capitalism for them to pigeonhole their users into their solutions through reducing solution accessibility to only their platforms.

Or maybe there is a more sensible, less tin foil hat reason for this and I am being short sighted. Either way Copilot Labs would be incredible to have in JetBrains IDEs.

**Comment by mchaniotakis** [comment]
When should we expect test generation for python?

**Comment by oluwatobiiloba** [comment]
I can't seem to run the tests generated. It returns "missing module" error
<img width="1131" alt="Screenshot 2023-01-31 at 11 54 36" src="https://user-images.githubusercontent.com/74077388/215741189-f874dc48-390f-40ca-be2b-020702749c8f.png">

**Reply by mattbrandman** [reply]
I had the same issue, we’re you able to resolve this?

**Comment by lucasmd** [comment]
Still no update about this on IntelliJ / Jetbrains ?

**Comment by lenguyenthanh** [comment]
please support other editors 🙏

**Comment by ruslaniv** [comment]
Could anybody please explain what `prompt` and `stop` are. I do not understand their purpose at all. What is the use case for these features?

**Reply by RobertoPelonara** [reply]
It has to do with "prompt design", the how you ask the AI to do the work basically. 
Some useful links here:
https://help.openai.com/en/articles/6654000-best-practices-for-prompt-engineering-with-openai-api
https://help.openai.com/en/articles/5072263-how-do-i-use-stop-sequences

**Comment by DominickVale** [comment]
Why are you choosing to leave out a good portion of developers that use jetbrains IDEs? Is the reason technical or...

**Reply by MCKevmeister** [reply]
Technically no access to that juicy juicy telemetry would be my guess

**Comment by MCKevmeister** [comment]
Please allow access to the tools offered by Github copilot labs outside of vs code, For example via api, terminal, other editors. Otherwise rename it for clarity to VS code labs

**Comment by kodeo** [comment]
Does GitHub Copilot Labs honor the same privacy setting "Allow GitHub to use my code snippets for product improvements" as the main product?

**Comment by MannyKayy** [comment]
I would like to try this with github copilot labs within visual studio enterprise.. is there a timeline for this?
